src/native/containers/dn-vector-t.h

```diff
@@ -0,0 +1,291 @@
+} DN_DEFINE_VECTOR_T_SYMBOL_NAME(name, sizes); \
+static inline void \
+DN_DEFINE_VECTOR_IT_T_SYMBOL_NAME(name, advance) (DN_DEFINE_VECTOR_IT_T_NAME(name) *it, ptrdiff_t n) \
+{ \
+	dn_vector_it_advance ((dn_vector_it_t *)it, n); \
+} \
+static inline DN_DEFINE_VECTOR_IT_T_NAME(name) \
+DN_DEFINE_VECTOR_IT_T_SYMBOL_NAME(name, next) (DN_DEFINE_VECTOR_IT_T_NAME(name) it) \
+{ \
+	DN_ASSERT ((it.it + 1) < UINT32_MAX); \
+	it.it++; \
+	return it; \
+} \
+static inline DN_DEFINE_VECTOR_IT_T_NAME(name) \
+DN_DEFINE_VECTOR_IT_T_SYMBOL_NAME(name, prev) (DN_DEFINE_VECTOR_IT_T_NAME(name) it) \
+{ \
+	DN_ASSERT (it.it > 0); \
+	it.it--; \
+	return it; \
+} \
+static inline DN_DEFINE_VECTOR_IT_T_NAME(name) \
+DN_DEFINE_VECTOR_IT_T_SYMBOL_NAME(name, next_n) (DN_DEFINE_VECTOR_IT_T_NAME(name) it, uint32_t n) \
+{ \
+	DN_ASSERT ((it.it + n) < UINT32_MAX); \
+	it.it += n; \
+	return it; \
+} \
+static inline DN_DEFINE_VECTOR_IT_T_NAME(name) \
+DN_DEFINE_VECTOR_IT_T_SYMBOL_NAME(name, prev_n) (DN_DEFINE_VECTOR_IT_T_NAME(name) it, uint32_t n) \
+{ \
+	DN_ASSERT (it.it >= n); \
+	it.it -= n; \
+	return it; \
+} \
+static inline type * \
+DN_DEFINE_VECTOR_IT_T_SYMBOL_NAME(name, data) (DN_DEFINE_VECTOR_IT_T_NAME(name) it, uint32_t n) \
+{ \
+	DN_ASSERT (it._internal._vector && it._internal._vector->data); \
+	return (type *)(it._internal._vector->data + n);\
+} \
+static inline DN_DEFINE_VECTOR_T_NAME(name) * \
+DN_DEFINE_VECTOR_T_SYMBOL_NAME(name, custom_alloc) (const DN_DEFINE_VECTOR_CUSTOM_ALLOC_PARAMS_T_SYMBOL_NAME(name) *params) \
+{ \
+	return (DN_DEFINE_VECTOR_T_NAME(name) *)_dn_vector_alloc_capacity (params->allocator, sizeof (type), params->capacity, params->attributes); \
+} \
+static inline DN_DEFINE_VECTOR_T_NAME(name) * \
+DN_DEFINE_VECTOR_T_SYMBOL_NAME(name, alloc) (void) \
+{ \
+	return (DN_DEFINE_VECTOR_T_NAME(name) *)_dn_vector_alloc (DN_DEFAULT_ALLOCATOR, sizeof (type), DN_VECTOR_ATTRIBUTE_INIT_MEMORY); \
+} \
+static inline void \
+DN_DEFINE_VECTOR_T_SYMBOL_NAME(name, free) (DN_DEFINE_VECTOR_T_NAME(name) *vector) \
+{ \
+	dn_vector_free ((dn_vector_t *)vector); \
+} \
+static inline void \
+DN_DEFINE_VECTOR_T_SYMBOL_NAME(name, custom_free) (DN_DEFINE_VECTOR_T_NAME(name) *vector, dn_vector_dispose_func_t dispose_func) \
+{ \
+	dn_vector_custom_free ((dn_vector_t *)vector, dispose_func); \
+} \
+static inline bool \
+DN_DEFINE_VECTOR_T_SYMBOL_NAME(name, custom_init) (DN_DEFINE_VECTOR_T_NAME(name) *vector, DN_DEFINE_VECTOR_CUSTOM_ALLOC_PARAMS_T_SYMBOL_NAME (name) *params) \
+{ \
+	return _dn_vector_init_capacity ((dn_vector_t *)vector, params->allocator, sizeof(type), params->capacity, params->attributes); \
+} \
+static inline bool \
+DN_DEFINE_VECTOR_T_SYMBOL_NAME(name, init) (DN_DEFINE_VECTOR_T_NAME(name) *vector) \
+{ \
+	return _dn_vector_init ((dn_vector_t *)vector, DN_DEFAULT_ALLOCATOR, sizeof (type), DN_VECTOR_ATTRIBUTE_INIT_MEMORY); \
```

Comment:
Again, let's make zero-initializing the customization structs be the default behavior. Negate the enum meaning:

```c
enum {
  DN_VECTOR_ATTRIBUTE_NO_INIT_MEMORY = 0x1
}
...
bool
DN_DEFINE_VECTOR_T_SYMBOL_NAME(name, custom_init) (DN_DEFINE_VECTOR_T_NAME(name) *vector, uint32_t element_size, DN_DEFINE_VECTOR_CUSTOM_ALLOC_PARAMS_T_SYMBOL_NAME (name) *params);

static inline bool
DN_DEFINE_VECTOR_T_SYMBOL_NAME(name, init) (DN_DEFINE_VECTOR_T_NAME(name) *vector)
{
  dn_vector_params_t params = {0,}; // DN_DEFALT_ALLOCATOR and no DN_VECTOR_ATTRIBUTE_NO_INIT_MEMORY
  return dn_vector_custom_init((dn_vector*)vector, sizeof(type), &params);
}
```